### PR TITLE
add ut for WorkloadRebalancer controller

### PR DESCRIPTION
**Pull request by chaosi-zju**
**DO NOT MERGE** until the following PR have been merged: 

* #4875

**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**:

add ut for WorkloadRebalancer controller

**Which issue(s) this PR fixes**:

Fixes part of #4840

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by RainbowMango** [comment]
/hold 
for #4875

**Comment by chaosi-zju** [comment]
cc @XiShanYongYe-Chang

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+	appsv1alpha1 "github.com/karmada-io/karmada/pkg/apis/apps/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+	"github.com/karmada-io/karmada/pkg/util/names"
+	"github.com/karmada-io/karmada/test/helper"
+)
+
+var (
+	now        = metav1.Now().Rfc3339Copy()
```
Comment (XiShanYongYe-Chang) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go:40]: Is there any particular reason to use Rfc3339Copy()?

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+			ObservedWorkloads: []appsv1alpha1.ObservedWorkload{
+				{
+					Workload: deploy1Obj,
+					Result:   appsv1alpha1.RebalanceSuccessful,
+				},
+			},
+		},
+	}
+	notFoundRebalancer = &appsv1alpha1.WorkloadRebalancer{
+		ObjectMeta: metav1.ObjectMeta{Name: "not-found-rebalancer", CreationTimestamp: now},
```
[the remark this answers is not in the document]
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go:77]: I will annotate it as `rebalancer-with-workloads-whose-binding-not-found`

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~
/lgtm
cc @RainbowMango

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &RebalancerController{
+				Client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).
+					WithObjects(tt.existObjects...).
+					WithStatusSubresource(tt.existObjsWithStatus...).Build(),
+			}
+			_, err := c.Reconcile(context.TODO(), tt.req)
+			// 1. check whether it has error
+			if (err != nil) != tt.wantErr {
```
[the remark this answers is not in the document]
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go:235]: This is not my self-created code style, it is auto-generated by `goland`. 

I do believe he has his own reasons for automatically generating such code.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+				t.Errorf("Reconcile() error = %v, wantErr %v", err, tt.wantErr)
+				return
```
[the remark this answers is not in the document]
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go]: this is also auto-generated, but I think yours makes more sense, done

**Comment by chaosi-zju** [comment]
is there any more comments?

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go:235]: You're saying that we should accept the code generated by Goland. It's not how it works.
Karmada always welcomes simple and readable code styles.

[the remark this answers is not in the document]
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller_test.go]: done

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**